### DISCUSSION TITLE
Unable to completely hide AIO behind the reverse proxy (Caddy)

### DISCUSSION THREAD

**Discussion by WDMdanila** [nextcloud/all-in-one#3720] [Questions] [answered]
Is there a way to configure caddy (and NC) in such a way that NC master container and Apache are actually fully behind it? What I want to achieve is that caddy docker is not in a host network so that I can point Caddy straight to the master container (which is technically done, but we still expose `8080` port for the master container as well as `11000` port for apache, which I would want to restrict to be reachable only through caddy)

Feels like such exposure of ports defeats the purpose of reverse proxy all together as services are reachable even without proxy.

Sorry for posting it here, did not know where to put such question and here seems like it would be the best place.

_Originally posted by @WDMdanila in https://github.com/nextcloud/all-in-one/discussions/575#discussioncomment-6122889_

**Comment by WDMdanila** [comment]
What I did right now looks like this:

My Caddyfile:
```Caddyfile
{
    email MY_EMAIL@ADDRESS
}

MY_DOMAIN {
	encode gzip
	reverse_proxy localhost:11000
}

https://MY_DOMAIN:8443 {
    reverse_proxy localhost:8080 {
        transport http {
            tls_insecure_skip_verify
        }
    }
}
```
docker-compose.yml
```compose
version: "3.8"

services:
  caddy:
    image: caddy:alpine
    restart: unless-stopped
    container_name: caddy
    network_mode: "host"
    volumes:
      - ./Caddyfile:/etc/caddy/Caddyfile
      - ./certs:/certs
      - ./config:/config
      - ./data:/data
      - ./sites:/srv

  nextcloud:
    image: nextcloud/all-in-one:latest
    restart: unless-stopped
    container_name: nextcloud-aio-mastercontainer
    ports:
      - "8080:8080"
    environment:
      - APACHE_PORT=11000
      - NEXTCLOUD_MOUNT=/mnt/nextcloud-nfs
    volumes:
      - nextcloud_aio_mastercontainer:/mnt/docker-aio-config
      - /var/run/docker.sock:/var/run/docker.sock:ro
    depends_on:
      - caddy

volumes:
  nextcloud_aio_mastercontainer:
    name: nextcloud_aio_mastercontainer
```

Is there a way to make it more isolated, so that Caddy is the only container actually exposing ports?

What I am thinking about is for example creating some networks which will be `external` for this compose, because those will be managed by AIO, but since they will be mentioned in a compose - caddy may be connected to those and thus be able to perform routing to them, without the need to expose container ports to host

**Comment by szaimen** [comment, marked as answer]
> Is there a way to make it more isolated, so that Caddy is the only container actually exposing ports?

No but see https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md#3-limit-the-access-to-the-apache-container

**Reply by WDMdanila** [reply]
Thank you!

Can this change be done after the initial deployment?

**Reply by szaimen** [reply]
> Can this change be done after the initial deployment?

Yes

**Reply by WDMdanila** [reply]
Should I firstly stop all containers from AIO interface? Sorry for bombarding you with questions - I am just very cautious about breaking anything :sweat_smile:

**Reply by WDMdanila** [reply]
Figured it out. Here is my compose:

```compose
version: "3.8"

volumes:
  nextcloud_aio_mastercontainer:
    name: nextcloud_aio_mastercontainer

networks:
  nextcloud:
    name: nextcloud
    driver: bridge
    ipam:
      config:
        - subnet: 172.18.0.0/16


services:
  caddy:
    image: caddy:alpine
    restart: unless-stopped
    container_name: caddy
    networks:
      nextcloud:
        ipv4_address: 172.18.0.2
    extra_hosts:
       - "host.docker.internal:host-gateway"
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - ./Caddyfile:/etc/caddy/Caddyfile
      - ./certs:/certs
      - ./config:/config
      - ./data:/data
      - ./sites:/srv

  nextcloud:
    image: nextcloud/all-in-one:latest
    restart: unless-stopped
    container_name: nextcloud-aio-mastercontainer
    networks:
      nextcloud:
        ipv4_address: 172.18.0.3
    environment:
      - APACHE_PORT=11000
      - APACHE_IP_BINDING=0.0.0.0
      - NEXTCLOUD_MOUNT=/mnt/nextcloud-nfs
    volumes:
      - nextcloud_aio_mastercontainer:/mnt/docker-aio-config
      - /var/run/docker.sock:/var/run/docker.sock:ro
    depends_on:
      - caddy
```

Caddyfile
```caddyfile
{
    email EMAIL@ADDRESS
}

nextcloud.DOMAIN {
	encode gzip
	reverse_proxy host.docker.internal:11000
}

aio.DOMAIN {
    reverse_proxy nextcloud:8080 {
        transport http {
            tls_insecure_skip_verify
        }
    }
}
```

To get it working had to also enable access to port `11000` in ufw via: `sudo ufw allow from 172.18.0.2 proto tcp to any port 11000`

Thanks @szaimen !
